iocore/net/UnixNetVConnection.cc

```diff
@@ -545,57 +545,61 @@ write_to_net_io(NetHandler *nh, UnixNetVConnection *vc, EThread *thread)
     vc->write.triggered = 0;
     write_signal_error(nh, vc, (int)-r);
     return;
-  } else {
-    int wbe_event = vc->write_buffer_empty_event; // save so we can clear if needed.
+  }
 
-    NET_SUM_DYN_STAT(net_write_bytes_stat, r);
+  int wbe_event = vc->write_buffer_empty_event; // save so we can clear if needed.
 
-    // Remove data from the buffer and signal continuation.
-    ink_assert(buf.reader()->read_avail() >= r);
-    buf.reader()->consume(r);
-    ink_assert(buf.reader()->read_avail() >= 0);
-    s->vio.ndone += r;
+  NET_SUM_DYN_STAT(net_write_bytes_stat, r);
 
-    // If the empty write buffer trap is set, clear it.
-    if (!(buf.reader()->is_read_avail_more_than(0)))
-      vc->write_buffer_empty_event = 0;
+  // Remove data from the buffer and signal continuation.
+  ink_assert(buf.reader()->read_avail() >= r);
+  buf.reader()->consume(r);
+  ink_assert(buf.reader()->read_avail() >= 0);
+  s->vio.ndone += r;
 
-    net_activity(vc, thread);
-    // If there are no more bytes to write, signal write complete,
-    ink_assert(ntodo >= 0);
-    if (s->vio.ntodo() <= 0) {
-      write_signal_done(VC_EVENT_WRITE_COMPLETE, nh, vc);
-      return;
-    } else if (signalled && (wbe_event != vc->write_buffer_empty_event)) {
+  // If the empty write buffer trap is set, clear it.
+  if (!(buf.reader()->is_read_avail_more_than(0)))
+    vc->write_buffer_empty_event = 0;
+
+  net_activity(vc, thread);
+  // If there are no more bytes to write, signal write complete,
+  ink_assert(ntodo >= 0);
+  int e = 0;
+  if (s->vio.ntodo() <= 0) {
+    write_signal_done(VC_EVENT_WRITE_COMPLETE, nh, vc);
+    return;
+  } else {
+    if (!signalled) {
+      e = VC_EVENT_WRITE_READY;
+    } else if (wbe_event != vc->write_buffer_empty_event) {
       // @a signalled means we won't send an event, and the event values differing means we
       // had a write buffer trap and cleared it, so we need to send it now.
-      if (write_signal_and_update(wbe_event, vc) != EVENT_CONT)
-        return;
-    } else if (!signalled) {
-      if (write_signal_and_update(VC_EVENT_WRITE_READY, vc) != EVENT_CONT) {
-        return;
-      }
-
-      // change of lock... don't look at shared variables!
-      if (lock.get_mutex() != s->vio.mutex.get()) {
-        write_reschedule(nh, vc);
-        return;
-      }
+      e = wbe_event;
     }
-
-    if (!buf.reader()->read_avail()) {
-      write_disable(nh, vc);
+  }
+  if (e) {
+    if (write_signal_and_update(e, vc) != EVENT_CONT) {
       return;
     }
 
-    if ((needs & EVENTIO_WRITE) == EVENTIO_WRITE) {
+    // change of lock... don't look at shared variables!
+    if (lock.get_mutex() != s->vio.mutex.get()) {
       write_reschedule(nh, vc);
+      return;
     }
-    if ((needs & EVENTIO_READ) == EVENTIO_READ) {
-      read_reschedule(nh, vc);
-    }
+  }
+
+  if (needs==0 && !buf.reader()->read_avail()) {
```

Comment:
Should use `read_avail_more_than(0)` here for performance as the exact number of bytes is irrelevant.